## Setting enviroment

In [1]:
# !pip install -q datasets==3.6.0

In [2]:
import datasets

print(datasets.__version__)

/home/mypc/Desktop/dlp/venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


3.6.0


## Que 1

In [3]:
from datasets import get_dataset_config_names

configs = get_dataset_config_names("ai4bharat/naamapadam")

print(configs)
print("Total configs:", len(configs))

['as', 'bn', 'gu', 'hi', 'kn', 'ml', 'mr', 'or', 'pa', 'ta', 'te']
Total configs: 11


In [4]:
from datasets import load_dataset

hi_ds = load_dataset("ai4bharat/naamapadam", "hi")

hi_ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 985787
    })
    test: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 867
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 13460
    })
})

In [5]:
from datasets import load_dataset

ds = load_dataset("ai4bharat/naamapadam", "ta", split='train+test+validation')

ds

Dataset({
    features: ['tokens', 'ner_tags'],
    num_rows: 501435
})

In [6]:
ds.cache_files

[{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'},
 {'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'},
 {'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]

In [7]:
ds.features

{'tokens': Sequence(feature=Value(dtype='string', id=None), length=-1, id=None),
 'ner_tags': Sequence(feature=ClassLabel(names=['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC'], id=None), length=-1, id=None)}

In [8]:
len(ds.features['ner_tags'].feature.names)

7

## Que 2

In [9]:
ds = load_dataset("ai4bharat/naamapadam", "ta")

ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 497882
    })
    test: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 758
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags'],
        num_rows: 2795
    })
})

In [10]:
ds.cache_files

{'train': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-train.arrow'}],
 'test': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-test.arrow'}],
 'validation': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/naamapadam-validation.arrow'}]}

## Que 3

`ds.cache_files`: represents the size of the downloaded dataset stored locally in the cache.

In [11]:
from datasets import load_dataset
import os

ds = load_dataset("ai4bharat/naamapadam", "ta")

total_size = 0

for split in ds.cache_files:
    for file_info in ds.cache_files[split]:
        size_mb = os.path.getsize(file_info["filename"]) / (1024 * 1024)
        total_size += size_mb

print(f"Total dataset size: {total_size:.2f} MB")

Total dataset size: 179.29 MB


### `load_dataset_builder()`

- Does not download or load the dataset itself
- Creates a dataset builder object containing metadata and download instructions
- Lets you inspect information about the dataset before downloading it

In [12]:
from datasets import load_dataset_builder

builder = load_dataset_builder("ai4bharat/naamapadam", "ta")


In [13]:
builder.info.features

{'tokens': Sequence(feature=Value(dtype='string', id=None), length=-1, id=None),
 'ner_tags': Sequence(feature=ClassLabel(names=['O', 'B-PER', 'I-PER', 'B-ORG', 'I-ORG', 'B-LOC', 'I-LOC'], id=None), length=-1, id=None)}

In [14]:
builder.info.description

'\n'

The total size of the files that need to be downloaded from the source.

- Think of it as, If you download a ZIP file from the internet: `dataset.zip  → 50 MB`
- then download_size is the size of that ZIP (or source files).

In [15]:
print("Download size:",
      builder.info.download_size / (1024*1024), "MB")

Download size: 46.37075233459473 MB


The size of the dataset after it has been processed and stored in Hugging Face's Arrow format.

- Downloaded files (50 MB)

        ↓

- Processed by datasets library

        ↓

- Arrow files (100 MB)

In [16]:
print("Dataset size:",
      builder.info.dataset_size / (1024*1024), "MB")

Dataset size: 179.84503650665283 MB


In [17]:
builder.info.splits

{'train': SplitInfo(name='train', num_bytes=187209743, num_examples=497882, shard_lengths=None, dataset_name='naamapadam'),
 'test': SplitInfo(name='test', num_bytes=320938, num_examples=758, shard_lengths=None, dataset_name='naamapadam'),
 'validation': SplitInfo(name='validation', num_bytes=1050508, num_examples=2795, shard_lengths=None, dataset_name='naamapadam')}

## Quick Summary

| Attribute | What it Represents | When Measured |
|------------|-------------------|----------------|
| `builder.info.download_size` | Size of the files downloaded from the dataset source (compressed/raw files) | Before processing |
| `builder.info.dataset_size` | Size of the processed dataset stored in Apache Arrow format | After processing |
| `os.path.getsize()` on `ds.cache_files` | Actual size of the cached dataset files stored on your machine | After download and processing |

### In Simple Terms

```text
Source Files (download_size)
            ↓
      Processing
            ↓
Arrow Dataset Files (dataset_size)
            ↓
Stored in Cache Directory
            ↓
Actual File Size on Disk (os.path.getsize)

## Que 4

In [18]:
ds['train'].num_rows

497882

## Que 5

In [19]:
ds['train'][0]

{'tokens': ['பைரவருக்கு',
  'தேய்பிறை',
  'அஷ்டமியில்',
  'விசேஷ',
  'அபிஷேக',
  'ஆராதனைகள்',
  'நடைபெறுகின்றன',
  '.'],
 'ner_tags': [1, 0, 0, 0, 0, 0, 0, 0]}

Creating `num tokens` column

In [20]:
def add_token_col(example):
    example['num tokens'] = len(example['tokens'])
    
    return example

In [21]:
ds = ds.map(add_token_col)

ds

DatasetDict({
    train: Dataset({
        features: ['tokens', 'ner_tags', 'num tokens'],
        num_rows: 497882
    })
    test: Dataset({
        features: ['tokens', 'ner_tags', 'num tokens'],
        num_rows: 758
    })
    validation: Dataset({
        features: ['tokens', 'ner_tags', 'num tokens'],
        num_rows: 2795
    })
})

In [22]:
ds['train'][0]['num tokens'] == len(ds['train'][0]['tokens'])


True

calculating the total number of columns in the entire dataset

In [23]:
total_tokens = 0

for split in ds:
    total_tokens += sum(ds[split]['num tokens'])
    
total_tokens/1_000_000

6.001876

## Que 6

In [24]:
ds.cache_files

{'train': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-339a5c9b5e072e5d.arrow'}],
 'test': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-4163f04c34e92fd8.arrow'}],
 'validation': [{'filename': '/home/mypc/.cache/huggingface/datasets/ai4bharat___naamapadam/ta/1.0.0/9d4f21ac57d11ed4f9ea64854fdc9f5618e61acc/cache-c572dfa9f0baec96.arrow'}]}

In [25]:
size = 0

for split in ds.cache_files:
    for file_info in ds.cache_files[split]:
        size += os.path.getsize(file_info['filename']) /(1024*1024)

print(f'Total_modified_size: {size} MB')

Total_modified_size: 183.1378173828125 MB


The original dataset occupied approximately 179.29 MB of disk space. After adding the "num tokens" column, the dataset size increased to approximately 183.14 MB. This corresponds to an increase of about 3.85 MB (≈2.15%). Therefore, the modified dataset did not increase the disk space requirement two-fold; it increased only marginally.

## Que 7

In [26]:
from datasets import concatenate_datasets

new_ds = concatenate_datasets(
    [
            ds['train'],
    ds['test'],
    ds['validation']
    ],
    axis=0
)

new_ds

Dataset({
    features: ['tokens', 'ner_tags', 'num tokens'],
    num_rows: 501435
})

In [27]:
def add_text_col(example):
    example['text'] = ' '.join(example['tokens'])
    
    return example

In [28]:
new_ds = new_ds.map(add_text_col)

new_ds[0]

{'tokens': ['பைரவருக்கு',
  'தேய்பிறை',
  'அஷ்டமியில்',
  'விசேஷ',
  'அபிஷேக',
  'ஆராதனைகள்',
  'நடைபெறுகின்றன',
  '.'],
 'ner_tags': [1, 0, 0, 0, 0, 0, 0, 0],
 'num tokens': 8,
 'text': 'பைரவருக்கு தேய்பிறை அஷ்டமியில் விசேஷ அபிஷேக ஆராதனைகள் நடைபெறுகின்றன .'}

In [29]:
ds = new_ds.remove_columns(['ner_tags', 'tokens'])

In [30]:
ds.column_names

['num tokens', 'text']

In [31]:
ds.num_rows

501435

In [32]:
ds

Dataset({
    features: ['num tokens', 'text'],
    num_rows: 501435
})

## Que 8


### Why Dataset Modifications Increase Disk Usage

Hugging Face datasets are **immutable**. Operations such as:

```python
ds = ds.map(...)
ds = ds.filter(...)
ds = ds.remove_columns(...)
ds = ds.rename_column(...)
```

do **not** modify the existing Apache Arrow file in place.

Instead, the library:

1. Reads the existing cached dataset.
2. Creates a **new Arrow cache file** containing the modified dataset.
3. Keeps the old cache file for reproducibility and caching purposes.

As a result, the cache directory may contain multiple versions of the dataset, causing disk usage to increase over time.

#### Example

Initial dataset:

```text
cache/
├── original.arrow
```

After:

```python
ds = ds.map(add_token_col)
```

Cache directory:

```text
cache/
├── original.arrow
├── modified.arrow
```

After another transformation:

```python
ds = ds.filter(...)
```

Cache directory:

```text
cache/
├── original.arrow
├── modified.arrow
├── filtered.arrow
```

#### Conclusion

Each dataset transformation typically creates a new cache file rather than modifying the existing one. Therefore, repeated operations such as `map()`, `filter()`, and `remove_columns()` can gradually increase the disk space consumed by the Hugging Face cache.

## Que 9

In [33]:
ds = ds.filter(lambda row: row['num tokens'] > 6)

Filter: 100%|██████████| 501435/501435 [00:01<00:00, 339236.20 examples/s]


In [43]:
ds

Dataset({
    features: ['num tokens', 'text'],
    num_rows: 332204
})

## Que 10

In [34]:
tamil_ds = load_dataset('ai4bharat/indic_glue', 'inltkh.ta', split='train+test+validation')

tamil_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 6684
})

In [35]:
tamil_ds = tamil_ds.filter(
    lambda row: len(row["text"].split()) > 6
)

In [36]:
tamil_ds

Dataset({
    features: ['text', 'label'],
    num_rows: 6213
})

In [37]:
tamil_ds = tamil_ds.remove_columns('label')

In [38]:
tamil_ds['text'][0]

'கே.வி.ஆனந்தே ட்விட்டரில் இதை அறிவித்துள்ளார். இந்தப் படத்துக்கு கேவ்மிக் ஆரி ஒளிப்பதிவு செய்ய, ஹாரிஸ் ஜெயராஜ் இசையமைக்கிறார். பட்டுக்கோட்டை பிரபாகர் வசனம் எழுத, கலை இயக்குநராக கிரண் பணியாற்றுகிறார். இந்தப் படத்தை லைகா புரொடக்\u200cஷன்ஸ் நிறுவனம் தயாரிக்கிறது.'

In [39]:
def add_col(col):
    col['num_tokens'] = len(col['text'].split(' '))
    
    return col

In [40]:
tamil_ds = tamil_ds.map(add_col)

Map: 100%|██████████| 6213/6213 [00:00<00:00, 20640.03 examples/s]


In [41]:
tamil_ds

Dataset({
    features: ['text', 'num_tokens'],
    num_rows: 6213
})

In [42]:
from datasets import interleave_datasets

id_data = interleave_datasets(
    [
        ds,
        tamil_ds,
    ],
    probabilities=[0.8, 0.2],
    seed = 42
)

id_data

Dataset({
    features: ['num tokens', 'text', 'num_tokens'],
    num_rows: 31178
})